athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchTypeUtils.java

```diff
@@ -0,0 +1,539 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.time.format.ResolverStyle;
+import java.util.ArrayList;
+import java.util.Map;
+
+/**
+ * This class has interfaces used for document field-values extraction after they are retrieved from an Elasticsearch
+ * instance. This includes field extractors and field writer factories using the field extractor framework.
+ */
+class ElasticsearchTypeUtils
+{
+    private static final Logger logger = LoggerFactory.getLogger(ElasticsearchTypeUtils.class);
+    private final ElasticsearchFieldResolver fieldResolver;
+
+    protected ElasticsearchTypeUtils()
+    {
+        this.fieldResolver = new ElasticsearchFieldResolver();
+    }
+
+    /**
+     * Create the appropriate field extractor used for extracting field values from a Document based on the field type.
+     * @param field is used to determine which extractor to generate based on the field type.
+     * @return a field extractor.
+     */
+    protected Extractor makeExtractor(Field field)
+    {
+        Types.MinorType fieldType = Types.getMinorTypeForArrowType(field.getType());
+
+        switch (fieldType) {
+            case VARCHAR:
+                return makeVarCharExtractor(field);
+            case BIGINT:
+                return makeBigIntExtractor(field);
+            case INT:
+                return makeIntExtractor(field);
+            case SMALLINT:
+                return makeSmallIntExtractor(field);
+            case TINYINT:
+                return makeTinyIntExtractor(field);
+            case FLOAT8:
+                return makeFloat8Extractor(field);
+            case FLOAT4:
+                return makeFloat4Extractor(field);
+            case DATEMILLI:
+                return makeDateMilliExtractor(field);
+            case BIT:
+                return makeBitExtractor(field);
+            default:
+                return null;
+        }
+    }
+
+    /**
+     * Create a VARCHAR field extractor to extract a string value from a Document. The Document value can be returned
+     * as a String or an ArrayList. For the latter, extract the first element only.
+     * @param field is used to determine which extractor to generate based on the field type.
+     * @return a field extractor.
+     */
+    private Extractor makeVarCharExtractor(Field field)
+    {
+        return (VarCharExtractor) (Object context, NullableVarCharHolder dst) ->
+        {
+            Object fieldValue = ((Map) context).get(field.getName());
+            dst.isSet = 1;
+            if (fieldValue instanceof String) {
+                dst.value = (String) fieldValue;
+            }
+            else if (fieldValue instanceof ArrayList) {
+                Object value = ((ArrayList) fieldValue).get(0);
+                if (value instanceof String) {
+                    dst.value = (String) value;
+                }
+                else {
+                    dst.isSet = 0;
```

Comment:
same comment as the next one.

Reply:
this is an indication that the field returned a null value (e.g. "field": null). In Elasticsearch the fields will be returned as the following class instances: String, Number, Boolean, ArrayList, HashMap,  and **null**.